diff --git a/crypto/evp/internal.h b/crypto/evp/internal.h
index 0a385c3..19594bc 100644
--- a/crypto/evp/internal.h
+++ b/crypto/evp/internal.h
@@ -286,6 +286,9 @@ inline auto GetDefaultEVPAlgorithms() {
       EVP_pkey_ed25519(),
       EVP_pkey_rsa(),
       EVP_pkey_x25519(),
+      EVP_pkey_ml_dsa_44(),
+      EVP_pkey_ml_dsa_65(),
+      EVP_pkey_ml_dsa_87(),
       // TODO(crbug.com/438761503): Remove DSA from this set, after callers that
       // need DSA pass in |EVP_pkey_dsa| explicitly.
       EVP_pkey_dsa(),
diff --git a/crypto/x509/algorithm.cc b/crypto/x509/algorithm.cc
index 6d06985..b73297a 100644
--- a/crypto/x509/algorithm.cc
+++ b/crypto/x509/algorithm.cc
@@ -57,9 +57,23 @@ int x509_digest_sign_algorithm(EVP_MD_CTX *ctx, X509_ALGOR *algor) {
     }
   }
 
-  if (EVP_PKEY_id(pkey) == EVP_PKEY_ED25519) {
-    return X509_ALGOR_set0(algor, OBJ_nid2obj(NID_ED25519), V_ASN1_UNDEF,
-                           nullptr);
+  ASN1_OBJECT *algo_obj = nullptr;
+  switch (EVP_PKEY_id(pkey)) {
+    case EVP_PKEY_ED25519:
+      algo_obj = OBJ_nid2obj(NID_ED25519);
+      break;
+    case EVP_PKEY_ML_DSA_44:
+      algo_obj = OBJ_nid2obj(NID_ML_DSA_44);
+      break;
+    case EVP_PKEY_ML_DSA_65:
+      algo_obj = OBJ_nid2obj(NID_ML_DSA_65);
+      break;
+    case EVP_PKEY_ML_DSA_87:
+      algo_obj = OBJ_nid2obj(NID_ML_DSA_87);
+      break;
+  }
+  if (algo_obj != nullptr) {
+    return X509_ALGOR_set0(algor, algo_obj, V_ASN1_UNDEF, nullptr);
   }
 
   // Default behavior: look up the OID for the algorithm/hash pair and encode
@@ -115,7 +129,8 @@ int x509_digest_verify_init(EVP_MD_CTX *ctx, const X509_ALGOR *sigalg,
     if (sigalg_nid == NID_rsassaPss) {
       return x509_rsa_pss_to_ctx(ctx, sigalg, pkey);
     }
-    if (sigalg_nid == NID_ED25519) {
+    if (sigalg_nid == NID_ED25519 || sigalg_nid == NID_ML_DSA_44 ||
+        sigalg_nid == NID_ML_DSA_65 || sigalg_nid == NID_ML_DSA_87) {
       if (sigalg->parameter != nullptr) {
         OPENSSL_PUT_ERROR(X509, X509_R_INVALID_PARAMETER);
         return 0;
diff --git a/crypto/obj/obj_xref.cc b/crypto/obj/obj_xref.cc
index a0ee1d9..d48de18 100644
--- a/crypto/obj/obj_xref.cc
+++ b/crypto/obj/obj_xref.cc
@@ -47,4 +47,7 @@ static const nid_triple kTriples[] = {
     // digest "undef" indicates the caller should handle this explicitly.
     {NID_rsassaPss, NID_undef, NID_rsaEncryption},
     {NID_ED25519, NID_undef, NID_ED25519},
+    {NID_ML_DSA_44, NID_undef, NID_ML_DSA_44},
+    {NID_ML_DSA_65, NID_undef, NID_ML_DSA_65},
+    {NID_ML_DSA_87, NID_undef, NID_ML_DSA_87},
 };
diff --git a/include/openssl/ssl.h b/include/openssl/ssl.h
index 7d15ce1..60bc3a6 100644
--- a/include/openssl/ssl.h
+++ b/include/openssl/ssl.h
@@ -1139,6 +1139,9 @@ OPENSSL_EXPORT int SSL_set_ocsp_response(SSL *ssl, const uint8_t *response,
 #define SSL_SIGN_RSA_PSS_RSAE_SHA384 0x0805
 #define SSL_SIGN_RSA_PSS_RSAE_SHA512 0x0806
 #define SSL_SIGN_ED25519 0x0807
+#define SSL_SIGN_ML_DSA_44 0x0904
+#define SSL_SIGN_ML_DSA_65 0x0905
+#define SSL_SIGN_ML_DSA_87 0x0906
 
 // SSL_SIGN_RSA_PKCS1_SHA256_LEGACY is a backport of RSASSA-PKCS1-v1_5 with
 // SHA-256 to TLS 1.3. It is disabled by default and only defined for client
diff --git a/ssl/extensions.cc b/ssl/extensions.cc
index a3b073b..1f022cd 100644
--- a/ssl/extensions.cc
+++ b/ssl/extensions.cc
@@ -284,6 +284,10 @@ static const uint16_t kVerifySignatureAlgorithms[] = {
 // algorithms for signing.
 static const uint16_t kSignSignatureAlgorithms[] = {
     // List our preferred algorithms first.
+    SSL_SIGN_ML_DSA_44,
+    SSL_SIGN_ML_DSA_65,
+    SSL_SIGN_ML_DSA_87,
+
     SSL_SIGN_ED25519,
     SSL_SIGN_ECDSA_SECP256R1_SHA256,
     SSL_SIGN_RSA_PSS_RSAE_SHA256,
diff --git a/ssl/ssl_privkey.cc b/ssl/ssl_privkey.cc
index f21c0d5..3190c5c 100644
--- a/ssl/ssl_privkey.cc
+++ b/ssl/ssl_privkey.cc
@@ -34,7 +34,8 @@ BSSL_NAMESPACE_BEGIN
 
 bool ssl_is_key_type_supported(int key_type) {
   return key_type == EVP_PKEY_RSA || key_type == EVP_PKEY_EC ||
-         key_type == EVP_PKEY_ED25519;
+         key_type == EVP_PKEY_ED25519 || key_type == EVP_PKEY_ML_DSA_44 ||
+         key_type == EVP_PKEY_ML_DSA_65 || key_type == EVP_PKEY_ML_DSA_87;
 }
 
 typedef struct {
@@ -98,6 +99,16 @@ static const SSL_SIGNATURE_ALGORITHM kSignatureAlgorithms[] = {
     {SSL_SIGN_ED25519, EVP_PKEY_ED25519, NID_undef, nullptr,
      /*is_rsa_pss=*/false, /*tls12_ok=*/true, /*tls13_ok=*/true,
      /*client_only=*/false},
+
+    {SSL_SIGN_ML_DSA_44, EVP_PKEY_ML_DSA_44, NID_undef, nullptr,
+     /*is_rsa_pss=*/false, /*tls12_ok=*/false, /*tls13_ok=*/true,
+     /*client_only=*/false},
+    {SSL_SIGN_ML_DSA_65, EVP_PKEY_ML_DSA_65, NID_undef, nullptr,
+     /*is_rsa_pss=*/false, /*tls12_ok=*/false, /*tls13_ok=*/true,
+     /*client_only=*/false},
+    {SSL_SIGN_ML_DSA_87, EVP_PKEY_ML_DSA_87, NID_undef, nullptr,
+     /*is_rsa_pss=*/false, /*tls12_ok=*/false, /*tls13_ok=*/true,
+     /*client_only=*/false},
 };
 
 static const SSL_SIGNATURE_ALGORITHM *get_signature_algorithm(uint16_t sigalg) {
@@ -115,8 +126,9 @@ bssl::UniquePtr<EVP_PKEY> ssl_parse_peer_subject_public_key_info(
   // code elimination, but for now we just specify every algorithm that might be
   // reachable from libssl.
   const EVP_PKEY_ALG *const algs[] = {
-      EVP_pkey_rsa(),     EVP_pkey_ec_p256(), EVP_pkey_ec_p384(),
-      EVP_pkey_ec_p521(), EVP_pkey_ed25519(),
+      EVP_pkey_rsa(),       EVP_pkey_ec_p256(),   EVP_pkey_ec_p384(),
+      EVP_pkey_ec_p521(),   EVP_pkey_ed25519(),   EVP_pkey_ml_dsa_44(),
+      EVP_pkey_ml_dsa_65(), EVP_pkey_ml_dsa_87(),
   };
   return bssl::UniquePtr<EVP_PKEY>(EVP_PKEY_from_subject_public_key_info(
       spki.data(), spki.size(), algs, std::size(algs)));
