diff --git a/include/openssl/ssl.h b/include/openssl/ssl.h
--- a/include/openssl/ssl.h
+++ b/include/openssl/ssl.h
@@ -1528,6 +1528,9 @@ DEFINE_CONST_STACK_OF(SSL_CIPHER)
 // Cipher 9F
 #define SSL_CIPHER_DHE_RSA_WITH_AES_256_GCM_SHA384 0x009F
 
+// Cipher CCAA
+#define SSL_CIPHER_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256 0xCCAA
+
 // SSL_get_cipher_by_value returns the structure representing a TLS cipher
 // suite based on its assigned number, or NULL if unknown. See
 // https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-parameters-4.
diff --git a/include/openssl/tls1.h b/include/openssl/tls1.h
--- a/include/openssl/tls1.h
+++ b/include/openssl/tls1.h
@@ -282,6 +282,10 @@ extern "C" {
 // Cipher 9F
 #define TLS1_TXT_DHE_RSA_WITH_AES_256_GCM_SHA384 "DHE-RSA-AES256-GCM-SHA384"
 
+// Cipher CCAA
+#define TLS1_TXT_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256 \
+  "DHE-RSA-CHACHA20-POLY1305"
+
 #ifdef __cplusplus
 }  // extern C
 #endif
diff --git a/ssl/ssl_cipher.cc b/ssl/ssl_cipher.cc
--- a/ssl/ssl_cipher.cc
+++ b/ssl/ssl_cipher.cc
@@ -493,6 +493,18 @@ static constexpr SSL_CIPHER kCiphers[] = {
         SSL_HANDSHAKE_MAC_SHA256,
     },
 
+    // Cipher CCAA
+    {
+        TLS1_TXT_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
+        "TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256",
+        SSL_CIPHER_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
+        SSL_kDHE,
+        SSL_aRSA_DECRYPT,
+        SSL_CHACHA20POLY1305,
+        SSL_AEAD,
+        SSL_HANDSHAKE_MAC_SHA256,
+    },
+
     // Cipher CCAB
     {
         TLS1_TXT_ECDHE_PSK_WITH_CHACHA20_POLY1305_SHA256,
@@ -1263,6 +1275,7 @@ bool ssl_create_cipher_list(UniquePtr<SSLCipherPreferenceList> *out_cipher_list,
       SSL_CIPHER_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,
       SSL_CIPHER_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
       SSL_CIPHER_ECDHE_PSK_WITH_CHACHA20_POLY1305_SHA256,
+      SSL_CIPHER_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
   };
   static const uint16_t kLegacyCiphers[] = {
       SSL_CIPHER_ECDHE_ECDSA_WITH_3DES_EDE_CBC_SHA,
