diff --git a/include/openssl/ssl.h b/include/openssl/ssl.h
--- a/include/openssl/ssl.h
+++ b/include/openssl/ssl.h
@@ -1145,6 +1145,8 @@
 #define SSL_SIGN_RSA_PSS_RSAE_SHA384 0x0805
 #define SSL_SIGN_RSA_PSS_RSAE_SHA512 0x0806
 #define SSL_SIGN_ED25519 0x0807
+#define SSL_SIGN_RSA_PKCS1_SHA224 0x0301
+#define SSL_SIGN_ECDSA_SHA224 0x0303
 #define SSL_SIGN_RSA_PSS_PSS_SHA256 0x0809
 #define SSL_SIGN_RSA_PSS_PSS_SHA384 0x080a
 #define SSL_SIGN_RSA_PSS_PSS_SHA512 0x080b
diff --git a/ssl/ssl_privkey.cc b/ssl/ssl_privkey.cc
--- a/ssl/ssl_privkey.cc
+++ b/ssl/ssl_privkey.cc
@@ -58,6 +58,9 @@
     {SSL_SIGN_RSA_PKCS1_SHA1, EVP_PKEY_RSA, NID_undef, &EVP_sha1,
      /*is_rsa_pss=*/false, /*tls12_ok=*/true, /*tls13_ok=*/false,
      /*client_only=*/false},
+    {SSL_SIGN_RSA_PKCS1_SHA224, EVP_PKEY_RSA, NID_undef, &EVP_sha224,
+     /*is_rsa_pss=*/false, /*tls12_ok=*/true, /*tls13_ok=*/false,
+     /*client_only=*/false},
     {SSL_SIGN_RSA_PKCS1_SHA256, EVP_PKEY_RSA, NID_undef, &EVP_sha256,
      /*is_rsa_pss=*/false, /*tls12_ok=*/true, /*tls13_ok=*/false,
      /*client_only=*/false},
@@ -97,6 +100,9 @@
     {SSL_SIGN_ECDSA_SHA1, EVP_PKEY_EC, NID_undef, &EVP_sha1,
      /*is_rsa_pss=*/false, /*tls12_ok=*/true, /*tls13_ok=*/false,
      /*client_only=*/false},
+    {SSL_SIGN_ECDSA_SHA224, EVP_PKEY_EC, NID_undef, &EVP_sha224,
+     /*is_rsa_pss=*/false, /*tls12_ok=*/true, /*tls13_ok=*/false,
+     /*client_only=*/false},
     {SSL_SIGN_ECDSA_SECP256R1_SHA256, EVP_PKEY_EC, NID_X9_62_prime256v1,
      &EVP_sha256, /*is_rsa_pss=*/false, /*tls12_ok=*/true, /*tls13_ok=*/true,
      /*client_only=*/false},
@@ -494,11 +500,13 @@
 static const SignatureAlgorithmName kSignatureAlgorithmNames[] = {
     {SSL_SIGN_RSA_PKCS1_MD5_SHA1, "rsa_pkcs1_md5_sha1"},
     {SSL_SIGN_RSA_PKCS1_SHA1, "rsa_pkcs1_sha1"},
+    {SSL_SIGN_RSA_PKCS1_SHA224, "rsa_pkcs1_sha224"},
     {SSL_SIGN_RSA_PKCS1_SHA256, "rsa_pkcs1_sha256"},
     {SSL_SIGN_RSA_PKCS1_SHA256_LEGACY, "rsa_pkcs1_sha256_legacy"},
     {SSL_SIGN_RSA_PKCS1_SHA384, "rsa_pkcs1_sha384"},
     {SSL_SIGN_RSA_PKCS1_SHA512, "rsa_pkcs1_sha512"},
     {SSL_SIGN_ECDSA_SHA1, "ecdsa_sha1"},
+    {SSL_SIGN_ECDSA_SHA224, "ecdsa_sha224"},
     {SSL_SIGN_ECDSA_SECP256R1_SHA256, "ecdsa_secp256r1_sha256"},
     {SSL_SIGN_ECDSA_SECP384R1_SHA384, "ecdsa_secp384r1_sha384"},
     {SSL_SIGN_ECDSA_SECP521R1_SHA512, "ecdsa_secp521r1_sha512"},
@@ -664,6 +672,7 @@
   uint16_t signature_algorithm;
 } kSignatureAlgorithmsMapping[] = {
     {EVP_PKEY_RSA, NID_sha1, SSL_SIGN_RSA_PKCS1_SHA1},
+    {EVP_PKEY_RSA, NID_sha224, SSL_SIGN_RSA_PKCS1_SHA224},
     {EVP_PKEY_RSA, NID_sha256, SSL_SIGN_RSA_PKCS1_SHA256},
     {EVP_PKEY_RSA, NID_sha384, SSL_SIGN_RSA_PKCS1_SHA384},
     {EVP_PKEY_RSA, NID_sha512, SSL_SIGN_RSA_PKCS1_SHA512},
@@ -671,6 +680,7 @@
     {EVP_PKEY_RSA_PSS, NID_sha384, SSL_SIGN_RSA_PSS_PSS_SHA384},
     {EVP_PKEY_RSA_PSS, NID_sha512, SSL_SIGN_RSA_PSS_PSS_SHA512},
     {EVP_PKEY_EC, NID_sha1, SSL_SIGN_ECDSA_SHA1},
+    {EVP_PKEY_EC, NID_sha224, SSL_SIGN_ECDSA_SHA224},
     {EVP_PKEY_EC, NID_sha256, SSL_SIGN_ECDSA_SECP256R1_SHA256},
     {EVP_PKEY_EC, NID_sha384, SSL_SIGN_ECDSA_SECP384R1_SHA384},
     {EVP_PKEY_EC, NID_sha512, SSL_SIGN_ECDSA_SECP521R1_SHA512},
